src/linux/ptrace_dumper.rs

```diff
@@ -160,8 +160,31 @@ impl PtraceDumper {
         ptrace::attach(pid).map_err(|e| AttachErr(child, e))?;
         loop {
             match wait::waitpid(pid, Some(wait::WaitPidFlag::__WALL)) {
-                Ok(_) => break,
-                Err(_e @ Errno::EINTR) => continue,
+                Ok(status) => {
+                    let wait::WaitStatus::Stopped(_, status) = status else {
+                        return Err(DumperError::WaitPidError(
+                            child,
+                            nix::errno::Errno::UnknownErrno,
+                        ));
+                    };
+
+                    // Any signal will stop the thread, make sure it is SIGSTOP. Otherwise, this
+                    // signal will be delivered after PTRACE_DETACH, and the thread will enter
+                    // the "T (stopped)" state.
+                    if status == nix::sys::signal::SIGSTOP {
+                        break;
+                    }
+
+                    // Signals other than SIGSTOP that are received need to be reinjected,
+                    // or they will otherwise get lost.
+                    // Note that the breakpad code doesn't do a detach, but that
+                    // feels like a bug...
```

Comment:
I was reading the `ptrace(2)` manpage and it turns out this is complicated to say the least. Detaching after the process woke up won't work because for it to work the thread must be stopped, so the page suggests re-stopping and re-injecting the signals again before detaching. So Breakpad is not wrong about not detaching but it's not doing the right thing either. I'm gonna check what gdb and lldb do to see if there's a way that doesn't require several nested loops of attaching/waiting/detaching to get it right. BTW the manpage seems to implicitly admit that this was a genuine design flaw.